dlt/common/plugins/plugins.py

```diff
@@ -0,0 +1,143 @@
+from dlt.common.typing import TDataItem
+from dlt.common.pipeline import SupportsPipeline
+from dlt.common.configuration.specs import ContainerInjectableContext
+from dlt.common.configuration.specs import configspec
+from functools import wraps
+from .reference import SupportsCallbackPlugin, Plugin, TSinglePluginArg, TPluginArg, CallbackPlugin
+from dlt.common.schema.exceptions import DataValidationError
+from importlib import import_module
+from .exceptions import UnknownPluginPathException
+from dlt.common.configuration.specs.base_configuration import BaseConfiguration
+import multiprocessing as mp
+from functools import wraps
+import threading
+import queue
+
+
+SAFE_SUFFIX = "_safe"
+
+
+def create_safe_version(f: TFun) -> TFun:
+    @wraps(f)
+    def _wrap(self: "PluginsContext", *args: Any, **kwargs: Any) -> Any:
+        # send message to shared queue if this is not the main instance
+        if self._plugins:
+            if not self._main or threading.main_thread() != threading.current_thread():
+                self._queue.put((f.__name__, args, kwargs))
+            else:
+                getattr(self, (f.__name__ + SAFE_SUFFIX))(*args, **kwargs)
+        return f(self, *args, **kwargs)
+
+    return _wrap  # type: ignore
+
+
+@configspec
+class PluginsContext(ContainerInjectableContext, SupportsCallbackPlugin):
+    def __init__(self, main: bool = True) -> None:
+        self._plugins: List[Plugin[BaseConfiguration]] = []
+        self._callback_plugins: List[CallbackPlugin[BaseConfiguration]] = []
+        self._initial_plugins: TPluginArg = []
+        self._main = main
+        self._queue: Optional[queue.Queue[Any]] = None
+
+    def _resolve_plugin(self, plugin: TSinglePluginArg) -> Plugin[BaseConfiguration]:
+        resolved_plugin: Plugin[BaseConfiguration] = None
+        if isinstance(plugin, str):
+            module_path, attr_name = plugin.rsplit(".", 1)
+            try:
+                module = import_module(module_path)
+            except ModuleNotFoundError as e:
+                raise UnknownPluginPathException(plugin) from e
+            try:
+                plugin = getattr(module, attr_name)
+            except AttributeError as e:
+                raise UnknownPluginPathException(str(plugin)) from e
+        if isinstance(plugin, type) and issubclass(plugin, Plugin):
+            resolved_plugin = plugin()
+        else:
+            raise TypeError(f"Plugin {plugin} is not a subclass of Plugin nor a plugin name string")
+        return resolved_plugin
+
+    # pickle support, send plugins config and queue to child process / threads
+    def __getstate__(self) -> Dict[str, Any]:
+        return {"plugins": self._initial_plugins, "queue": self._queue}
+
+    def __setstate__(self, d: Dict[str, Any]) -> None:
+        self.__init__(False)  # type: ignore[misc]
+        self.setup_plugins(d["plugins"])
+        self._queue = d["queue"]
+
+    def process_queue(self) -> None:
+        assert self._main
```

Comment:
Maybe adding a message to assert will be helpful for debugging, something like `Expected to run in the main thread`